lib/at_monitor/at_monitor.c

```diff
@@ -31,6 +31,17 @@ static void at_monitor_dispatch(const char *notif)
 	struct at_notif_fifo *at_notif;
 	size_t sz_needed;
 
+	/* Dispatch to monitors in ISR, if any.
+	 * There might be non-ISR monitors that are interested
+	 * in the same notification, so copy it to the heap afterwards regardless.
+	 */
+	STRUCT_SECTION_FOREACH(at_monitor_isr_entry, e) {
+		if (!e->paused && (e->filter == ANY || strstr(notif, e->filter))) {
+			LOG_DBG("Dispatching to %p (ISR)", e->handler);
+			e->handler(notif);
+		}
+	}
+
 	sz_needed = sizeof(struct at_notif_fifo) + strlen(notif) + sizeof(char);
 
 	at_notif = k_heap_alloc(&at_monitor_heap, sz_needed, K_NO_WAIT);
```

Comment:
The log below this line is a bit inconvenient in SMS use case when we are using AT_MONITOR_ISR.

This is how it looks in modem_shell when a message is received:
```
Time:   22-01-04 12:14:43
Text:   '123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123'
Length: 153
Concatenated short messages: ref_number=122, msg 1/4
00:01:13.053,283] <wrn> at_monitor: No heap space for incoming notification: +CMT: "+358408031652",159
0791534874894370440C91534880306125000022104021413480A00500037A0401~

```

Reply:
This is related to question 1 from @lemrey in the first comment: https://github.com/nrfconnect/sdk-nrf/pull/6402#issue-1081025419

There's no deferred monitor for the +CMT notification, but the implementation still attempts to make a copy of it. I think it's good to have a warning trace about the failed alloc. The (in this case unnecessary) warning could be removed by adding a loop to check whether there's a deferred monitor which matches the notification before allocating memory. Maybe the check should be added, because this is a normal use case for AT_MONITOR_ISR. What do you @lemrey and @trantanen think?

Reply:
Well, it's somewhat related to question #1. I think it's ok to reserve heap for all notifications (even though some may not have a monitor) but the warning log is really inconvenient.
@lemrey wants to avoid the processing of iterating through all the monitors but that would be the easiest way to avoid the extra log (and at the same time unnecessary memory allocation). @lemrey can comment on the performance penalty of this change.

Reply:
I have updated the code so that it copies only monitored notifications. I don't think there should be any noticeable impact as the number of monitors is small.

Reply:
as long as there is at least one `ANY` monitor, all notifications will be copied, so you will still see the log. I think we should keep it there. It's a warning and not an error anyway.

Reply:
I think this is fine, because that `ANY` monitor would not get that notification when memory allocation fails.

Reply:
If modem_shell user doesn't request all notifications to be printed to the shell (this is OFF by default), there wouldn't be ANY monitor unless lte_lc wouldn't have one. But in practice, when we use at_monitor library, we also use lte_lc library which will cause all notifications to be requested.

Looking at lte_lc, it could rather easily be modified to avoid an unnecessary ANY monitor.
But from at_monitor perspective, this is the best we can do although it'll generate questions from modem_shell users.